### ISSUE TITLE
The possibility to erase successfuly scraped metrices

### ISSUE THREAD

**Issue by mjf** [google/mtail#84]
Hello.

I would like to count on *evergrowing* set of metric indices (i.e. relay hosts from mail logs or similar case)
but it's clear that free memory on the host is likely to be exhausted after some time which makes *mtail*
less useful in general.

A solution to this "issue" may be to have **the possibility to erase metrices that were successfuly
scraped**.

This may be achieved by either allowing marking them with special keyword when declaring them (i.e. ``[hidden] session counter <metric> ...``) and erasing these *automatically* after successful scrape **or** by extending the ``del`` functionality by extending the syntax with something like ``scrape del <metric>`` (ugly?) **or** with sort of section like *awk* does have (*awk* has ``BEGIN { ... }`` and ``END { ... }`` sections, *mtail* can have at least the ``END { ... }`` section to place anything right after every successful scrape, i.e. ``END { del <metric> }``).

The maintenance of the metrices with evergrowing indices would be moved to a scraper this way, and on the scraper-side data retention policy could be set, some calculations done, new metrices created, etc.

**Comment by jaqx0r** [comment]
It is a neat idea, a signal to garbage collection that the datum is stale.

The problem with removing a labelset and datum from the export is that it confuses the collector, instead of seeing a timeseries stable, it sees a timeseries disappear (i.e. there are missing samples, instead of constant.)  I am not sure what, e.g. Prometheus would do with a rate calculation in the absence of data points if one appears in the future.  So adding a feature that lets us garbage collect any labelset would need warnings around it that it has the potential to confuse a collector.  I'd like to understand more and write about the behaviour of other tools in the mtail documentation before making any feature like this not experimental.

The problem with running a block of code after scrape is that we support both push and pull based metric collection.  However that does mean we could clean up immediately after a metric push in those cases, i.e. wait for the latest delay necessary before executing the code.  Determining that no Prometheus or other pull collector is defined may require a commandline flag to disable it so that garbage collection can be done as soon as all the push is done.

I'd rename the block to SCRAPE or POSTSCRAPE or something, though, as mtail programs never END in the awk sense.

**Comment by mjf** [comment]
@jaqx0r Well, the ``SCRAPE`` and ``POSTSCRAPE`` look just good-enough (although I do not understand the ``SCRAPE`` idea). What exactly would one use the ``SCRAPE{}`` block for (from your point of view)?

To be precise, my exact idea was this: ``BEGIN{}``, collect, collect, collect, ..., **successful scrape**, ``END{}``, **new line arrives in the log**, ``BEGIN{}``, collect, ..., and so on (where "collect" meas "process" the line). I mean, this is slighly different, right? ``BEGIN{}`` means "hey, there's new work, let's begin" and ``END{}`` means "hey, we're done for now, the end, let's wait for a new work". 😄 

I was considering the ``BEGIN{}`` block to start at the exact point when first new line arrives in the logs after successful scrape, before it get parsed or passed to the interpreter (or how to call it), and the ``END{}`` block right after a successful scrape. Also the time "gap" between both blocks could be measured if one wanted (no real or important use case for that than just to have it complete).

**Comment by jaqx0r** [comment]
Now I'm wondering if just a TTL of 1 day on any metric is good enough, and we close out the ones that don't exist anymore after that -- that's long enough to scraep, and we can make the TTL a flag so you can make, say, your mtail expiry TTL 2 times the valye of a prometheus scrape interval if you want.

**Comment by jaqx0r** [comment]
Setting a timeout like that seems like a good approach to the problem.  It
is not magic, and it informs the programmer that there's some time after
which a monitoring system won't see the values.

Thanks for the suggestion!

**Comment by jwalzer** [comment]
of course, another approach to my problem would be to think about #112 - because all the different state-labels that I'm tracking could be used as free-form annotations, that don't create a separate metric (as is currently the case)

Then I would only need to create an Metric by the "peer_id" and have the state as an annotation, thats exported to prometheus. But when the state changes, the old metric is no longer exported as a stale one, but only the peer and its new state is exported

Does make sense?
